### ISSUE TITLE
SIP 002 - Use Native ES6 Modules Without Bundling

### ISSUE THREAD

**Issue by Luis-Fernando-Molina** [Superalgos/Improvement-Proposals#5]
This is the discussion thread for SIP 002:

https://github.com/Superalgos/Improvement-Proposals/blob/master/SIP%20002%20-%20ESM-no-webpack-no-babel.md

**Comment by Luis-Fernando-Molina** [comment]
Not an expert on these topics, but the proposal makes sense to me.

**Comment by bearcanrun** [comment]
I agree in principle, but there's some major refactoring, or really, setting some standards first and then applying it to the whole code base to make this happen. ...and then expecting all contributions to apply it.

Frontend specific holds most uncertainty for me in this issue: With Webpack 4/ Babel 7, the bundler learning curve has really gone down — and for the most part will be pre-setup in the repo. Specifically, the use of hot module reloading(for style dev is there an alternative?) and dynamic imports is something that most people learning and developing with React are used to. Granted, the JS ecosystem landscape is still maturing and maybe by the time we have the contribution power/time-for-rebuilding-everything-from-ground-zero this will be a moot point. 

Just to respond directly to the factors:
- *Page Loading*: Agreed that code splitting is a great optimization must, so if anything, we get an award for not pre-optimizing (our bundle size is a case study of non-optimized decentralized collaboration...). The labor is the same with or without bundler. ECMAScript dynamic import() is still proposal, so still need Babel and don't know of/how-to-do any other alternatives.
- *Debugging*: Comments, not arguments - Perhaps code splitting is already optimization enough, but we may find the need to have as highly-optimized and streamlined a product as possible — especially given the possible future complexity of interconnecting pieces. Ideally we would have a strong TDD/regression tests/deployment policies in place to maximize bug mitigation in the first place... 
- *Licenses*: This is an issue, but may best be dealt with education (licenses for many people are just plain confusing) and as a part of the collaborative culture in the first place. `Before adding new library, check license` Then also, hopefully caught in code reviews. No doubt we already have license auditing issues and why might be easiest/best to refactor from ground zero. 
- *Code Complexity*: "one recognized standard" ... 🤔. I was about to majorly digress, when I remembered we are _just_ talking about "Native ES6 Modules Without Bundling." Phew...

**Comment by isysd-mirror** [comment]
I am not sure it is actually such major refactoring that is required. In my previous experiences with this transition, it is mostly a matter of ripping out unnecessary complexity rather than re-organizing what is left.

Specifically, in regards to hot reloading, this is a bundling problem in search of a solution. When modules are loaded one by one, there is no pre-processing, and therefore no hot reloading required. Honestly, I mainly made this proposal because I was sick of waiting for the hot reload...

I believe the main ways we'd need to refactor would be:

 + Change to absolute or relative imports from name (i.e. `../../axios` instead of `axios`)
 + Add a single nodejs entry file ala esm package publishing. (5m per package)
 + Link built versions of used libraries instead of listing them as package.json dependencies, which may require minor scripting to run `npm run build` and/or `rollup` for each.

**Comment by isysd-mirror** [comment]
Out of curiosity, I confirmed that yes dynamic imports are still a proposal. They are also supported in every major browser, covering 78% of browsers.

https://caniuse.com/#search=dynamic%20import

**Comment by isysd-mirror** [comment]
Just thought of a potential hangup with this: JSX is *not* supported by most browsers and engines. For JSX, pre-processing will remain a requirement. Considering how heavily react based we currently are, this might be a significant challenge.

**Comment by bearcanrun** [comment]
We are pretty React heavy. 

Hot reload on Master App is poorly implemented. Should only be used on lower scopes to alleviate total bundle rebuild. With code-splitting should improve further.  I've also done the component development more "black-boxed" using Storybook/TDD which makes things much easier and would be a suggested implementation flow once we get our UX dev setup in place.

Not too distant future is the hard look at refactor/build-from-scratch to a more UX-friendly app(s). Modules packages will be less view/more components. Also multiple smaller apps for specific tasks/stakeholders instead of a monolithic "master" app.

**Comment by isysd-mirror** [comment]
OK, that all makes sense, and logically this discussion can be tabled until then. Thanks for the feedback, @bearcanrun!

**Comment by isysd-mirror** [comment]
I created a boilerplate for isomorphic modules, which enforces the standards suggested in this SIP.

https://github.com/isysd-mirror/iso-module-boilerplate
